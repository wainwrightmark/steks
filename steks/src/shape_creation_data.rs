use crate::prelude::*;
use bevy::{ecs::system::EntityCommands, render::color};
use bevy_prototype_lyon::prelude::*;
use rand::{rngs::StdRng, seq::SliceRandom, Rng};

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct ShapeCreationData {
    pub shape: &'static GameShape,
    pub location: Option<Location>,
    pub state: ShapeState,
    pub velocity: Option<Velocity>,
    pub modifiers: ShapeModifiers,
    pub id: Option<u32>,
}

pub fn add_components(state: &ShapeState, ec: &mut EntityCommands) {
    if state == &ShapeState::Void {
        ec.insert(VoidShape { highlighted: false });
        ec.insert(Sensor {});
        ec.insert(WallSensor);
        ec.insert(ActiveEvents::COLLISION_EVENTS);
    } else if state == &ShapeState::Fixed {
        ec.insert(FixedShape);
    }
}

pub fn remove_components(state: &ShapeState, ec: &mut EntityCommands) {
    if state != &ShapeState::Void {
        ec.remove::<VoidShape>();
        ec.remove::<Sensor>();
        ec.remove::<ActiveEvents>();
        ec.remove::<WallSensor>();
    }

    if state != &ShapeState::Fixed {
        ec.remove::<FixedShape>();
    }
}

pub fn spawn_children(
    cb: &mut ChildBuilder,
    shape: &GameShape,
    state: ShapeState,
    transform: &Transform,
) {
    cb.spawn_empty()
        .insert(Shadow)
        .insert(shape.body.get_shape_bundle(SHAPE_SIZE * ZOOM_LEVEL))
        .insert(Transform {
            translation: Vec3::new(0., 0., 10.),
            ..Default::default()
        })
        .insert(Visibility::Hidden)
        .insert(Stroke {
            color: Color::BLACK,
            options: StrokeOptions::default().with_line_width(ZOOM_LEVEL),
        });

    // if state == ShapeState::Void {
    //     let transform = Transform {
    //         rotation: transform.rotation.conjugate(),
    //         scale: PADLOCK_SCALE,
    //         translation: Vec3::Z * 50.0,
    //     };

    //     let path = GeometryBuilder::build_as(&shapes::SvgPathShape {
    //         svg_path_string: SKULL_OUTLINE.to_owned(),
    //         svg_doc_size_in_px: SVG_DOC_SIZE.to_owned(),
    //     });

    //     cb.spawn(ShapeBundle {
    //         path,
    //         ..Default::default()
    //     })
    //     .insert(Fill {
    //         options: FillOptions::DEFAULT,
    //         color: WARN_COLOR,
    //     })
    //     .insert(transform)
    //     .insert(Visibility::Inherited);
    // }

    if state == ShapeState::Fixed {
        let transform = Transform {
            rotation: transform.rotation.conjugate(),
            scale: PADLOCK_SCALE,
            translation: Vec3::Z * 50.0,
        };

        let path = GeometryBuilder::build_as(&shapes::SvgPathShape {
            svg_path_string: PLAIN_PADLOCK_OUTLINE.to_owned(),
            svg_doc_size_in_px: SVG_DOC_SIZE.to_owned(),
        });

        cb.spawn(ShapeBundle {
            path,
            ..Default::default()
        })
        .insert(Fill {
            options: FillOptions::DEFAULT,
            color: Color::BLACK,
        })
        .insert(transform)
        .insert(Visibility::Inherited);
    }
}

impl ShapeCreationData {
    pub fn fill(&self) -> Fill {
        self.state.fill().unwrap_or_else(|| self.shape.fill())
    }

    pub fn stroke(&self) -> Stroke {
        self.state.stroke().unwrap_or_else(|| {
            self.modifiers.stroke().unwrap_or_else(|| Stroke {
                color: color::Color::NONE,
                options: StrokeOptions::DEFAULT.with_line_width(0.0),
            })
        })
    }

    pub fn velocity_component(&self) -> Velocity {
        self.velocity.unwrap_or_default()
    }
}

impl From<EncodableShape> for ShapeCreationData {
    fn from(value: EncodableShape) -> Self {
        let EncodableShape {
            shape,
            location,
            state,
            modifiers,
        } = value;

        Self {
            shape,
            location: Some(location),
            state: state,
            velocity: None,
            modifiers,
            id: None,
        }
    }
}

impl ShapeCreationData {
    pub fn by_name(s: &str) -> Option<Self> {
        GameShape::by_name(s).map(|shape| Self {
            shape,
            location: None,
            state: ShapeState::Normal,
            velocity: Some(Default::default()),
            modifiers: ShapeModifiers::Normal,
            id: None,
        })
    }

    pub fn with_location(mut self, position: Vec2, angle: f32) -> Self {
        self.location = Some(Location { position, angle });
        self
    }

    pub fn lock(mut self) -> Self {
        self.state = ShapeState::Locked;
        self
    }

    pub fn with_velocity(mut self, velocity: Velocity) -> Self {
        self.velocity = Some(velocity);
        self
    }

    pub fn with_random_velocity(mut self) -> Self {
        self.velocity = None;
        self
    }

    pub fn from_seed(seed: u64) -> Self {
        let mut shape_rng: StdRng = rand::SeedableRng::seed_from_u64(seed);
        Self::random(&mut shape_rng)
    }

    pub fn random<R: Rng>(shape_rng: &mut R) -> Self {
        let shape = ALL_SHAPES.choose(shape_rng).unwrap();

        Self {
            shape,
            location: None,
            state: ShapeState::Normal,
            velocity: Some(Default::default()),
            modifiers: ShapeModifiers::Normal,
            id: None,
        }
    }
}
