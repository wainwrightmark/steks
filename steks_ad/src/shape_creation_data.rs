use crate::prelude::*;
use bevy::{ecs::system::EntityCommands, render::color};
use bevy_prototype_lyon::prelude::*;

#[derive(Debug, Clone, Copy, PartialEq, Event)]
pub struct ShapeCreationData {
    pub shape: &'static GameShape,
    pub location: Option<Location>,
    pub state: ShapeState,
    pub velocity: Option<Velocity>,
    pub modifiers: ShapeModifiers,
    pub id: Option<u32>,
    pub color: Option<Color>,
    pub stage: ShapeStage,
}

pub fn add_components(state: &ShapeState, ec: &mut EntityCommands) {
    if state == &ShapeState::Void {
        ec.insert(VoidShape { highlighted: false });
        ec.insert(Sensor {});
        ec.insert(WallSensor);
        ec.insert(ActiveEvents::COLLISION_EVENTS);
    } else if state == &ShapeState::Fixed {
        ec.insert(FixedShape);
    }
}

pub fn remove_components(state: &ShapeState, ec: &mut EntityCommands) {
    if state != &ShapeState::Void {
        ec.remove::<VoidShape>();
        ec.remove::<Sensor>();
        ec.remove::<ActiveEvents>();
        ec.remove::<WallSensor>();
    }

    if state != &ShapeState::Fixed {
        ec.remove::<FixedShape>();
    }
}

pub fn spawn_children(
    cb: &mut ChildBuilder,
    shape: &GameShape,
    state: ShapeState,
    transform: &Transform,
) {
    cb.spawn_empty()
        .insert(Shadow)
        .insert(shape.body.get_shape_bundle(SHAPE_SIZE * ZOOM_LEVEL))
        .insert(Transform {
            translation: Vec3::new(0., 0., 10.),
            ..Default::default()
        })
        .insert(Visibility::Hidden)
        .insert(Stroke {
            color: state.shadow_stroke(),
            options: StrokeOptions::default().with_line_width(ZOOM_LEVEL),
        });

    if state == ShapeState::Fixed {
        let transform = Transform {
            rotation: transform.rotation.conjugate(),
            scale: PADLOCK_SCALE,
            translation: Vec3::Z * 50.0,
        };

        let path = GeometryBuilder::build_as(&shapes::SvgPathShape {
            svg_path_string: PLAIN_PADLOCK_OUTLINE.to_owned(),
            svg_doc_size_in_px: SVG_DOC_SIZE.to_owned(),
        });

        cb.spawn(ShapeBundle {
            path,
            ..Default::default()
        })
        .insert(Fill {
            options: FillOptions::DEFAULT,
            color: Color::BLACK,
        })
        .insert(transform)
        .insert(Visibility::Inherited);
    }
}

impl ShapeCreationData {
    pub fn fill(&self, high_contrast: bool) -> Fill {
        if let Some(color) = self.color {
            return Fill {
                color,
                options: FillOptions::DEFAULT,
            };
        }

        self.state
            .fill()
            .unwrap_or_else(|| self.shape.fill(high_contrast))
    }

    pub fn stroke(&self, high_contrast: bool) -> Stroke {
        self.state.stroke().unwrap_or_else(|| {
            self.modifiers
                .stroke(high_contrast)
                .unwrap_or_else(|| Stroke {
                    color: color::Color::NONE,
                    options: StrokeOptions::DEFAULT.with_line_width(0.0),
                })
        })
    }

    pub fn velocity_component(&self) -> Velocity {
        self.velocity.unwrap_or_default()
    }
}

impl ShapeCreationData {
    pub fn from_encodable(value: EncodableShape, stage: ShapeStage) -> Self {
        let EncodableShape {
            shape,
            location,
            state,
            modifiers,
        } = value;

        Self {
            shape,
            location: Some(location),
            state,
            velocity: None,
            modifiers,
            id: None,
            color: None,
            stage,
        }
    }

    pub fn from_shape_creation(shape_creation: ShapeCreation, stage: ShapeStage) -> Self {
        let mut fixed_location: Location = Default::default();
        let mut fl_set = false;
        if let Some(x) = shape_creation.x {
            fixed_location.position.x = x;
            fl_set = true;
        }
        if let Some(y) = shape_creation.y {
            fixed_location.position.y = y;
            fl_set = true;
        }
        if let Some(r) = shape_creation.r {
            fixed_location.angle = r * std::f32::consts::TAU;
            fl_set = true;
        }

        let fixed_location = fl_set.then_some(fixed_location);

        let velocity = match shape_creation.state {
            ShapeState::Locked | ShapeState::Fixed | ShapeState::Void => Some(Default::default()),
            ShapeState::Normal => {
                if shape_creation.vel_x.is_some() || shape_creation.vel_y.is_some() {
                    Some(Velocity {
                        linvel: Vec2 {
                            x: shape_creation.vel_x.unwrap_or_default(),
                            y: shape_creation.vel_y.unwrap_or_default(),
                        },
                        angvel: Default::default(),
                    })
                } else {
                    None
                }
            }
        };

        ShapeCreationData {
            shape: shape_creation.shape.into(),
            location: fixed_location,
            state: shape_creation.state,
            velocity,
            modifiers: shape_creation.modifiers,
            id: shape_creation.id,
            color: shape_creation.color.map(|(r, g, b)| Color::rgb_u8(r, g, b)),
            stage,
        }
    }

    pub fn from_shape_index(shape_index: ShapeIndex, stage: ShapeStage) -> Self {
        Self {
            shape: shape_index.into(),
            location: None,
            state: ShapeState::Normal,
            velocity: Some(Default::default()),
            modifiers: ShapeModifiers::Normal,
            id: None,
            color: None,
            stage,
        }
    }

    pub fn with_location(mut self, position: Vec2, angle: f32) -> Self {
        self.location = Some(Location { position, angle });
        self
    }

    pub fn lock(mut self) -> Self {
        self.state = ShapeState::Locked;
        self
    }

    pub fn with_velocity(mut self, velocity: Velocity) -> Self {
        self.velocity = Some(velocity);
        self
    }

    pub fn with_random_velocity(mut self) -> Self {
        self.velocity = None;
        self
    }
}
