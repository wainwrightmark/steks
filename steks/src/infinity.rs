use itertools::Itertools;
use rand::{rngs::StdRng, SeedableRng};

use crate::prelude::*;

pub fn get_initial_shapes(seed: u64) -> Vec<ShapeIndex> {
    let mut rng: StdRng = StdRng::seed_from_u64(seed);
    let mut shapes: Vec<ShapeIndex> = vec![];
    for _ in 0..INFINITE_MODE_STARTING_SHAPES {
        shapes.push(ShapeIndex::random_no_circle(&mut rng));
    }

    shapes
}

pub fn get_all_shapes(seed: u64, total_shapes: usize) -> Vec<ShapeCreationData> {
    let mut collected = get_initial_shapes(seed);
    let mut stage = ShapeStage(0);
    let mut results: Vec<ShapeCreationData> = collected
        .iter()
        .map(|x| ShapeCreationData::from_shape_index(*x, stage))
        .clone()
        .collect_vec();
    collected.sort();

    while collected.len() < total_shapes {
        stage.increment();
        let mut hash: u64 = 0;
        for s in collected.iter() {
            hash  = hash.wrapping_mul(97).wrapping_add(s.0 as u64)
        }
        let next_shape_index = ShapeIndex::from_seed_no_circle(hash);

        results.push(ShapeCreationData::from_shape_index(next_shape_index, stage));
        collected.push(next_shape_index);
        collected.sort();
    }

    results
}

#[cfg(test)]
mod tests {
    use steks_common::prelude::INFINITE_MODE_STARTING_SHAPES;

    use crate::shape_creation_data::ShapeCreationData;

    use super::get_all_shapes;

    #[test_case::case(123)]
    #[test_case::case(456)]
    #[test_case::case(789)]
    pub fn check_shapes_sequence(seed: u64) {
        let shapes = get_all_shapes(seed, 50);

        let indices: Vec<_> = shapes.iter().map(|x| x.shape.index.0).collect();

        insta::assert_debug_snapshot!(indices);
    }

    #[test]
    pub fn check_shapes_consistency(){

        let mut prev_shapes = get_all_shapes(123, INFINITE_MODE_STARTING_SHAPES);
        for x in 1..=10{
            let new_shapes = get_all_shapes(123, INFINITE_MODE_STARTING_SHAPES + x);
            let prefix : Vec<ShapeCreationData>= new_shapes.iter().take(prev_shapes.len()).cloned().collect();
            assert_eq!(prefix, prev_shapes);
            prev_shapes = new_shapes;
        }
    }
}
