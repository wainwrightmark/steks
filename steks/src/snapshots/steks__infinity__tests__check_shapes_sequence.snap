---
source: steks/src/infinity.rs
expression: indices
---
[
    20,
    4,
    15,
    17,
    10,
    19,
    9,
    7,
    10,
    2,
    11,
    5,
    19,
    12,
    18,
    17,
    8,
    8,
    15,
    2,
    10,
    15,
    6,
    1,
    10,
    12,
    19,
    13,
    13,
    2,
    5,
    20,
    13,
    19,
    11,
    10,
    15,
    18,
    19,
    11,
    4,
    9,
    16,
    9,
    4,
    6,
    10,
    15,
    18,
    13,
]
